File: ui/src/services/store.ts
import { Base, Drive } from "deta";
import { nanoid } from "nanoid";
import { camelizeKeys, decamelizeKeys } from "humps";
import { ContentType, IBookmark } from "../interfaces";
import { FetchOptions } from "deta/dist/types/types/base/request";
import { DetaType } from "deta/dist/types/types/basic";

interface BookmarkResponse {
    key: string;
    content_type: string;
    snipped: string;
    metadata: Record<string, any>;
}

class Store {
    private static base = Base("bookmarks");
    private static drive = Drive("bookmarks");

    static config = {
        base: "api",
        add: "add",
        addFile: "add/file",
        fetch: "fetch",
        download: "download",
        delete: "delete"
    }

    static buildFilename(key: string, metadata: Record<string, any>): string {
        const fileExtension = metadata["file_extension"];
        return fileExtension ? `${key}${fileExtension}` : key;
    }

    static async addFile(key: string, file: File) {
        const item = await this.base.get(key) as unknown as BookmarkResponse;

        // file store not needed if base item doesn't exist or if content type is not file-like
        if ((!item) || (item["content_type"] == ContentType.TEXT)) {
            return
        }

        const filename = this.buildFilename(key, item["metadata"]);
        const buffer = new Uint8Array(await file.arrayBuffer());

        try {
            await this.drive.put(filename, { data: buffer, contentType: item["content_type"] });
            return key;
        } catch (err: any) {
            throw Error("Sorry, something is wrong! Couldn't sync your bookmark's file. Please try refreshing.");
        }
    }

    static async add(bookmark: IBookmark): Promise<IBookmark> {
        try {
            await this.base.put(decamelizeKeys(bookmark) as DetaType);
            return bookmark;
        } catch (error: any) {

            throw Error("Sorry, somehting is wrong! Couldn't sync your bookmark. Please try refreshing.");
        }
    }

    static async fetch(last?: string): Promise<IBookmark[]> {
        try {
            const options: FetchOptions = last ? { last } : {};
            const response = await this.base.fetch({}, options);
            const bookmarks = await response.items;
            return bookmarks.map((bookmark) => camelizeKeys(bookmark) as IBookmark);
        } catch (error: any) {
            throw Error("Sorry, something is wrong! Please try refreshing.")
        }
    }

    static async delete(key: string): Promise<string> {
        const item = await this.base.get(key) as unknown as BookmarkResponse;

        if (!item) {
            return key
        }

        try {
            await this.base.delete(key);
        } catch (err) {
            throw Error("Sorry, something is wrong! Couldn't delete your bookmark.");
        }

        if (item["content_type"] == ContentType.TEXT) {
            return key
        }

        const filename = await this.buildFilename(key, item["metadata"]);

        try {
            await this.drive.delete(filename);
            return key
        } catch (error: any) {
            throw Error("Sorry, something is wrong! Couldn't delete your bookmark's image.")
        }
    }

    static key() {
        const maxDateNowValue = 8.64e15;

        const id = nanoid(10);
        const timestamp = maxDateNowValue - Date.now();

        return `${timestamp}_${id}`
    }
}

export default Store;

File: api/store.py
import os
from deta import Deta
from fastapi import UploadFile
from models import Bookmark, ContentType
import util

deta = Deta()

"""
Store
"""


class Store:
    def __init__(self):
        self._base = deta.Base("bookmarks")
        self._drive = deta.Drive("bookmarks")

    def add(self, bookmark: Bookmark):
        item = {
            "snippet": bookmark.snippet,
            "key": bookmark.key or util.key(),
            "content_type": bookmark.content_type,
            "metadata": bookmark.metadata,
        }

        try:
            self._base.put(item)
        except Exception as e:
            raise Exception(
                "An error occurred while storing a bookmark in base."
            ) from e

        return item

    def __build_filename(self, key: str, metadata: dict):
        file_extension = metadata["file_extension"]

        if file_extension:
            return f"{key}{file_extension}"

        return key

    def add_file(self, key: str, file: UploadFile):
        item = self._base.get(key)

        # file store not needed if base item doesn't exist or if content type is not file-like
        if (not item) or (item["content_type"] == ContentType.TEXT):
            return None

        filename = self.__build_filename(key, item["metadata"])

        try:
            self._drive.put(filename, data=file.file, content_type=file.content_type)
        except Exception as e:
            raise Exception(
                "An error occurred while storing a bookmark in drive."
            ) from e

        return key

    def delete(self, key: str):
        item = self._base.get(key)

        if not item:
            return key

        content_type = item["content_type"]

        try:
            self._base.delete(key)
        except Exception as e:
            raise Exception(
                "An error occurred while deleting a bookmark from base."
            ) from e

        if content_type == ContentType.FILE or content_type == ContentType.IMAGE:
            try:
                filename = self.__build_filename(key, item["metadata"])
                self._drive.delete(filename)
            except Exception as e:
                raise Exception(
                    "An error occurred while deleting a bookmark from drive."
                ) from e

        return key

    def delete_all(self):
        bookmarks = self.fetch_all()
        for bookmark in bookmarks:
            self.delete(bookmark["key"])

    def fetch(self, last=None, limit=50, query={}):
        res = self._base.fetch(query=query, limit=limit, last=last)
        return res.items

    def fetch_all(self):
        res = self._base.fetch()
        all_bookmarks = res.items

        while res.last:
            res = self._base.fetch(last=res.last)
            all_bookmarks += res.items
        return all_bookmarks

    def download_file(self, name: str):
        res = self._drive.get(name)
        return res

    def cleanup(self):
        res = self._drive.list()
        files = res["names"]
        while res.get("paging") and res["paging"].get("last"):
            res = self._drive.list(last=res["paging"].get("last"))
            files += res["names"]

        files_to_delete = []
        for file in files:
            key, _ = os.path.splitext(file)
            if not self._base.get(key):
                files_to_delete.append(file)

        if len(files_to_delete):
            print(f"Deleting #{files_to_delete}")
            self._drive.delete_many(files_to_delete)
